### PR TITLE
Generalized FlagSet class

### LINKED ISSUE #12152: Generalized `FlagSet` class
Related to https://github.com/ethereum/solidity/pull/11944#discussion_r711333725.

In #11944 we added the [`DebugInfoSelection`](https://github.com/ethereum/solidity/blob/develop/liblangutil/DebugInfoSelection.h) class that represents a set of flags for selecting various debug info components. It provides helpers for parsing these flags from CLI and Standard JSON input and allows iterating over the flags so that they do not have to be repeated every time (which is very error-prone).

The task here is to implement a generalized class representing such sets of flags and then refactor `DebugInfoSelection` and other classes that work in a similar way to use it:
- [`SolverChoice`](https://github.com/ethereum/solidity/blob/v0.8.7/libsmtutil/SolverInterface.h#L39-L107)
- [`ModelCheckerEngine`](https://github.com/ethereum/solidity/blob/v0.8.7/libsolidity/formal/ModelCheckerSettings.h#L115-L142)
- [`ModelCheckerTargets`](https://github.com/ethereum/solidity/blob/v0.8.7/libsolidity/formal/ModelCheckerSettings.h#L92-L113)
- [`CombinedJsonRequests`](https://github.com/ethereum/solidity/blob/v0.8.7/solc/CommandLineParser.h#L76-L98)
- [`CompilerOutputs`](https://github.com/ethereum/solidity/blob/v0.8.7/solc/CommandLineParser.h#L54-L74)
- [`OptimiserSettings`](https://github.com/ethereum/solidity/blob/v0.8.10/libsolidity/interface/OptimiserSettings.h) (this one has some non-boolean fields so it might not be a perfect fit for this).
- [`FunctionType::Options`](https://github.com/ethereum/solidity/blob/v0.8.13/libsolidity/ast/Types.h#L1259-L1289)

### Implementation notes.
I think that the best way to accomplish this is to make the class a template and have each specific set of flags derive from it. Only the fields themselves and the function that enumerates them would need to be defined in the derived class.

Another approach would be to make it a variadic template and define fields via template parameters. I'm not sure if this is doable though.

### PR DESCRIPTION
**Pull request by a1is43ras4**
Fixes #12152 

- Added macro [DEFINE_METHOD_CHECK](https://github.com/justunkn/solidity/blob/issue_12152/Generalized_FlagSet_class/libsolutil/pch.h#L26) to generate meta functions to check existance of method in a given class
- Added [FlagSet](https://github.com/justunkn/solidity/blob/issue_12152/Generalized_FlagSet_class/libsolutil/FlagSet.h#L54) generalized class implementation
- All of listed classes except [OptimiserSettings](https://github.com/ethereum/solidity/blob/v0.8.10/libsolidity/interface/OptimiserSettings.h) are refactored by inheriting class  
[FlagSet](https://github.com/justunkn/solidity/blob/issue_12152/Generalized_FlagSet_class/libsolutil/FlagSet.h#L54)

### CODE REVIEW

File: libsolidity/formal/ModelCheckerSettings.h [hunk 1]
```diff
@@ -160,15 +176,9 @@ struct ModelCheckerSettings
...
-			divModNoSlacks == _other.divModNoSlacks &&
-			engine == _other.engine &&
-			invariants == _other.invariants &&
-			showUnproved == _other.showUnproved &&
-			solvers == _other.solvers &&
-			targets == _other.targets &&
-			timeout == _other.timeout;
+		return contracts == _other.contracts && divModNoSlacks == _other.divModNoSlacks && engine == _other.engine
+			   && invariants == _other.invariants && showUnproved == _other.showUnproved && solvers == _other.solvers
+			   && targets == _other.targets && timeout == _other.timeout;
```
Comment (cameel) [inline libsolidity/formal/ModelCheckerSettings.h]: This PR has a lot of unnecessary reformatting that just makes it harder to see actual changes. Can you please remove formatting only-changes and make sure that the code you're adding matches the style of the code around it?

**Comment by cameel** [comment]
If this PR is meant to be a complete solution for #12152, please include `Fixes #12152` in PR description. With that the PR number in the title is not necessary.

File: libsolidity/formal/ModelCheckerSettings.h (hunk 1 above)
Comment (a1is43ras4) [reply libsolidity/formal/ModelCheckerSettings.h]: @cameel, removed extra reformatting and fixed code style. I had some troubles with formatting in my ide. That's why there were a lots of unnecessary reformatting. But, now it seems to be fixed and works ok

**Comment by a1is43ras4** [comment]
@cameel btw, do you have any suggestions why the ci checking of the PR is still not completed?

**Comment by cameel** [comment]
@justunkn Have you seen [my response to your question on #solidity-dev](https://gitter.im/ethereum/solidity-dev?at=625436cee9cb3c52ae5ca51d)?

File: libsolutil/pch.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+#include <type_traits>
+
+/**
+ * Generates a meta function `trait` to check existence of `method` with signature $sign in the given class `U`
+ *
+ */
+#define DEFINE_METHOD_CHECK(trait, method, sign)						                \
```
Comment (cameel) [inline libsolutil/pch.h]: Not sure if there's a point in having a macro for this if it's used only once. Why not just put it directly in `FlagSet.h`?

File: libsolutil/pch.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/*
```
Comment (cameel) [inline libsolutil/pch.h]: `pch.h` is not a very clear name for a header file. Is that meant to be a "precompiled header" or something? We already have some headers for common stuff (names starting with `Common`) but this is very specific and should have a descriptive name, e.g. `MethodCheck.h`.

File: libsolutil/pch.h (hunk 2 above)
Comment (cameel) [inline libsolutil/pch.h]: Please don't use abbreviations in names.
```suggestion
#define DEFINE_METHOD_CHECK(trait, method, signature)						                \
```

File: libsolutil/pch.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Generates a meta function `trait` to check existence of `method` with signature $sign in the given class `U`
+ *
+ */
+#define DEFINE_METHOD_CHECK(trait, method, sign)						                \
+	template<typename U>                    						                    \
+	class trait										                                    \
+	{											                                        \
+	private:                                						                    \
```
Comment (cameel) [inline libsolutil/pch.h]: You seem to have a mix of tabs and spaces here. Please use tabs only for indentation, spaces in all other places.

File: libsolutil/FlagSet.h [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+#include <libsolutil/StringUtils.h>
+
+#include <boost/algorithm/string/trim.hpp>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/map.hpp>
+#include <range/v3/view/split.hpp>
+
+#include "StringUtils.h"
```
Comment (cameel) [inline libsolutil/FlagSet.h]: This is already included above. Also, please note we always use the `<>` style for includes and paths relative to repo root.

File: libsolutil/FlagSet.h [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+		for (auto const& [flagName, member]: Derived::flagMap())
+			if (flag == member)
+				return flagName;
+
+		solAssert(false, "");
+	}
+
+	static optional<Derived> fromFlags(vector<string> const& _flagNames, bool _acceptWildcards = false)
+	{
+		solAssert(Derived::flagMap().count("*") == 0, "");
```
Comment (cameel) [inline libsolutil/FlagSet.h:116]: We have recently updated this macro to make the message optional so you can drop the argument when writing new code, like here:
```suggestion
		solAssert(Derived::flagMap().count("*") == 0);
```

File: test/libyul/ObjectParser.cpp [hunk 7]
```diff
@@ -74,7 +74,7 @@ pair<bool, ErrorList> parse(string const& _source)
 	return {false, {}};
 }
 
-optional<Error> parseAndReturnFirstError(string const& _source, bool _allowWarningsAndInfos = true)
+std::optional<Error> parseAndReturnFirstError(string const& _source, bool _allowWarningsAndInfos = true)
```
Comment (cameel) [inline test/libyul/ObjectParser.cpp]: We do not add these in .cpp files where we have `using namespace std`:
```suggestion
optional<Error> parseAndReturnFirstError(string const& _source, bool _allowWarningsAndInfos = true)
```

File: libsolutil/CMakeLists.txt [hunk 8]
```diff
@@ -1,44 +1,46 @@
...
-	UTF8.h
-	vector_ref.h
-	Views.h
-	Visitor.h
-	Whiskers.cpp
-	Whiskers.h
+        Algorithms.h
+        AnsiColorized.h
+        Assertions.h
+        Common.h
```
Comment (cameel) [inline libsolutil/CMakeLists.txt:5]: This looks as if the whole list was rewritten but it's just that the indentation has been changed here from spaces to tabs. Please keep the original style.

File: libsolidity/formal/ModelCheckerSettings.h [hunk 9]
```diff
@@ -115,29 +99,45 @@ struct ModelCheckerInvariants
...
-	static ModelCheckerTargets Default() { return *fromString("default"); }
-	/// Adds all targets, including underflow and overflow.
-	static ModelCheckerTargets All() { return *fromString("all"); }
+	ModelCheckerTargets() = default;
 
-	static std::optional<ModelCheckerTargets> fromString(std::string const& _targets);
+	ModelCheckerTargets(size_t nFlags, ...);
 
-	bool has(VerificationTargetType _type) const { return targets.count(_type); }
+	[[nodiscard]] bool has(VerificationTargetType _type) const;
```
Comment (cameel) [inline libsolidity/formal/ModelCheckerSettings.h]: Why is `[[nodiscard]]` needed here? The result is just a boolean.

File: libsolidity/formal/ModelCheckerSettings.cpp [hunk 10]
```diff
@@ -82,36 +82,84 @@ map<TargetType, string> const ModelCheckerTargets::targetTypeToString{
...
-	return true;
+	}
+}
+
+bool ModelCheckerTargets::has(VerificationTargetType _type) const
+{
+	switch (_type)
+	{
+	case VerificationTargetType::ConstantCondition:
+		return constantCondition;
```
Comment (cameel) [inline libsolidity/formal/ModelCheckerSettings.cpp:143]: Maybe we could add translation to/from an enum as a feature to the `FlagSet`? It's a bit unfortunate that with the new class this part actually got more repetitive.

**Review by cameel — COMMENTED** [review]
Here's some feedback from a quick look at the code. I only skimmed over it for now so it's mostly superficial things. I'll take another look once these are fixed.

Overall this looks pretty nice and apart from styling/following our conventions I haven't seen any big problems so far.

**Comment by a1is43ras4** [comment]
@cameel i've missed it. but now i see.

**Review by nishant-sachdeva — COMMENTED** [review]
@justunkn  Are you still active on this PR ? Or shall we go ahead and takeover ?

**Comment by cameel** [comment]
@justunkn if you want to resume working on it later, feel free to reopen. For the time being I'm closing this since we're flooded with PRs and this not a critical roadmap feature (it's just a "nice to have") so it does not make much sense for us to take over and finish it ourselves.

**Pull request closed without merging**